staging/src/k8s.io/client-go/tools/cache/mutation_cache.go

```diff
@@ -260,3 +289,67 @@ func (a etcdObjectVersioner) CompareResourceVersion(lhs, rhs runtime.Object) int
 
 	return 1
 }
+
+type generationComparator struct {}
+
+func (c generationComparator) generation(obj runtime.Object) (int64, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return 0, err
+	}
+	return accessor.GetGeneration(), nil
+}
+
+// ObservedGenerationAccesor knows how to surface all generations of the object that have been
+// observed by controllers, whether that is at a top-level or in specific conditions.
+type ObservedGenerationAccesor interface {
+	// ObservedGenerations surfaces all generations of the object that have been observed
+	ObservedGenerations() []int64
+}
+
+func (c generationComparator) observedGeneration(obj runtime.Object) (int64, error) {
+	accessor, ok := obj.(ObservedGenerationAccesor)
+	if !ok {
+		return 0, fmt.Errorf("%T does not expose observed generations", obj)
+	}
+	var latestObservedGeneration int64
+	for _, observedGeneration := range accessor.ObservedGenerations() {
+		if observedGeneration > latestObservedGeneration {
+			latestObservedGeneration = observedGeneration
+		}
+	}
```

Comment:
trying to reason about whether this is the right behaviour to take the maximum. If there are multiple controllers, each responsible for a subset of observed generations, I think each want to compare with "its own" observed generations.

Imagine controller A and B, writing `status.observedGeneration{A,B}`. `ObjectMeta.generation` changes to 42, with `status.observedGeneration{A,B} == 41`. Controller A observes 42, updates the object status with `status.observedGenerationA = 42`. Controller B observes 42, but finds that `observedGeneration(obj) == 42` already. So B ignores the generation 42.

Reply:
Good point, each controller with their own cache will need a controller-specific getter for the observed generation, which makes me think that even more that a per-use implementation of the getter is what we want. I'll refactor, thanks.